libsysbox/syscont/spec.go

```diff
@@ -690,24 +692,68 @@ func sysMgrSetupMounts(mgr *sysbox.Mgr, spec *specs.Spec, uidShiftRootfs bool) e
 	}
 
 	// Otherwise, add the special dir to the list of mounts that we will request
-	// sysbox-mgr to setup
+	// sysbox-mgr to setup. Sysbox-mgr will setup host dirs to back the mounts in the
+	// request list; it will also send us any other mounts it needs.
+
+	rootPath, err := filepath.Abs(spec.Root.Path)
+	if err != nil {
+		return err
+	}
+
 	reqList := []ipcLib.MountReqInfo{}
 	for dest, kind := range specialDir {
+
+		// Check if the special dir requires uid shifting or not
+		path := filepath.Join(rootPath, dest)
+		shiftUids := false
+
+		fi, err := os.Stat(path)
+
+		if err == nil {
+
+			// If the special dir exists within the container's rootfs, then uid
+			// shifting is required when it's owned by root:root.
+
+			st, _ := fi.Sys().(*syscall.Stat_t)
+
+			if st.Uid != st.Gid {
+				return fmt.Errorf("container rootfs has special dir %s with non-matching uid & gid: %u %u", path, st.Uid, st.Gid)
+			}
+
+			if st.Uid != 0 && st.Uid != uid {
```

Comment:
Don't we need this same exact if-statement for st.Gid?

Reply:
Yes you are right; I added it.